odl/solvers/functional/derivatives.py

```diff
@@ -0,0 +1,326 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Utilities for computing the gradient and Hessian of functionals."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+from odl.solvers.functional.functional import Functional
+from odl.operator.operator import Operator
+from odl.space.base_ntuples import FnBase
+
+
+__all__ = ('NumericalDerivative', 'NumericalGradient',)
+
+
+class NumericalDerivative(Operator):
+
+    """The derivative of a operator by forward differences.
+
+    See Also
+    --------
+    NumericalGradient : Compute gradient of a functional
+    """
+
+    def __init__(self, operator, point, method='forward', step=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        functional : `Operator`
+            The operator whose derivative should be computed. Its domain and
+            range must be `FnBase` spaces.
+        point : ``functional.domain`` `element-like`
+            The point to compute the derivative in.
+        method : {'backward', 'forward', 'central'}
+            The method to use to compute the derivative.
+        step : `None` or float
+            The step length used in the derivative computation.
+            `None` is interpreted as selecting the step according to the dtype
+            of the space. ``step = 1e-8`` for ``float64`` spaces and
+            ``step = 1e-4`` for ``float32`` spaces.
```

Comment:
Alternatively, be less specific here with numbers, you could just state the implementation, namely the square root of the data type precision.